content/module2/3-http.md

```diff
@@ -0,0 +1,145 @@
+```
+
+### Headers
+
+This  is essentially metadata of the request.
+
+```
+Host: www.example.com  
+User-Agent: Mozilla/5.0  
+Accept: text/html  
+```
+These are standard agreed headers that HTTP can send, but are not mandatory. This is what they mean:
+* **Host** – Specifies the target server.  - We'll need to study DNS in order to understand this a bit better.
+* **User-Agent** – Identifies the client (browser, bot, etc.).
+* **Accept** – Indicates the content type the client can handle. - on browser you typicaly accept `html`, but WebServices communication typically happens in `json`.
+
+To understand headers in detail, visit [this link](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers). This `Mozilla` website is a great reference for HTTP standards and deep diving into the protocol.
+
+Because HTTP is based on TCP, TCP headers are also sent, but they should not to be confused with HTTP Headers which aren't present in TCP.
+
+Which is why you can only route with IP using a TCP Load Balancer and can't route based on hostname, because Host is an HTTP header.
+
+## Body - Optional
+
+Not all HTTP request have body. In fact, depending on the method, some **cannot** have a body.
+
+```json
+{
+  "username": "john_doe",
+  "password": "securepassword"
+}
+```
+This will contain data sent to the server (e.g., form submissions, JSON payloads).
+
+## HTTP Response Structure
+
+### Status Line
+
+A server’s HTTP response consists of:
+
+```
+HTTP/1.1 200 OK
+```
+* **Protocol Version** – HTTP/1.1 or HTTP/2.*
+* **Status Code** – Indicates success (2XX), error (4XX, 5XX), or redirection (3XX).
+* **Status Message** – Human-readable response (e.g., OK, Not Found).
+
+### Headers
+
+```
+Content-Type: text/html  
+Content-Length: 3456  
+```
+
+These headers need to be present if there is a body, they specify the `type` of its content as well as its length in `bytes`.
+
+### Body - Optional
+
+Contains the requested content (HTML, JSON, an image, etc.).
+
+```json
+{
+  "id": 3,
+  "name": "Kalchatzi",
+  "active": true
+}
+```
+
+## HTTP Verbs/Actions
+
+When you make an HTTP request you need to specify the action (verb). These are the ones available:
+
+* GET – Retrieves data (read-only).
+* POST – Sends data to create a resource (e.g., form submission).
+* PUT – Updates or replaces a resource.
+* PATCH – Partially updates a resource.
+* DELETE – Removes a resource.
+* HEAD – Like GET, but without a response body (for metadata).
+* OPTIONS – Retrieves allowed HTTP methods for a resource.
+
+## HTTP Status Codes
+* 1xx (Informational): Processing, e.g., 100 Continue.
+* 2xx (Success): Request completed, e.g., 200 OK, 201 Created.
+* 3xx (Redirection): Resource moved, e.g., 301 Moved Permanently, 302 Found.
+* 4xx (Client Errors): Bad requests, e.g., 400 Bad Request, 404 Not Found.
+* 5xx (Server Errors): Server issues, e.g., 500 Internal Server Error, 503 Service Unavailable.
```

Comment:
I would link to this for a complete list: https://en.wikipedia.org/wiki/List_of_HTTP_status_codes